core/src/main/java/org/elasticsearch/action/bulk/TransportShardBulkAction.java

```diff
@@ -495,13 +496,20 @@ static boolean shouldExecuteReplicaItem(final BulkItemRequest request, final int
         // Verify that there are no more mappings that need to be applied. If there are failures, a
         // ReplicationOperation.RetryOnPrimaryException is thrown.
         final Engine.Index operation;
-        try {
-            operation = prepareIndexOperationOnPrimary(request, primary);
-            mappingUpdater.verifyMappings(operation, primary.shardId());
-        } catch (MapperParsingException | IllegalStateException e) {
-            // there was an error in parsing the document that was not because
-            // of pending mapping updates, so return a failure for the result
-            return new Engine.IndexResult(e, request.version());
+        final Mapping mappingUpdate = preUpdateOperation.parsedDoc().dynamicMappingsUpdate();
```

Comment:
Above this, we don't look at whether we need to do the mapping update but let it to `updateMappingsIfNeeded` to decide. I think it will be clearer if we capture this as a boolean and use it to call `updateMappingsIfNeeded` (now renamed to `updateMappings`).